lang/src/dispatcher.rs

```diff
@@ -15,229 +15,111 @@
-pub type EmptyDispatchList = UnreachableDispatcher;
-
-impl EmptyDispatchList {
-    /// Creates a new dispatch list.
-    #[inline(always)]
-    pub fn empty() -> EmptyDispatchList {
-        UnreachableDispatcher
-    }
-}
-
-impl<Head, Rest> PushDispatcher for DispatchList<Head, Rest> {
-    /// Pushes another dispatcher onto the list.
-    #[inline(always)]
-    fn push<D>(self, _dispatcher: D) -> DispatchList<D, Self> {
-        Default::default()
-    }
-}
-
-impl<Head, Rest> Dispatch for DispatchList<Head, Rest>
-where
-    Head: Dispatch + FnSelector,
-    Rest: Dispatch,
-{
-    #[inline(always)]
-    fn dispatch(data: &CallData) -> Result<()> {
-        if <Head as FnSelector>::SELECTOR == data.selector() {
-            <Head as Dispatch>::dispatch(data)
-        } else {
-            <Rest as Dispatch>::dispatch(data)
-        }
-    }
-}
-
-/// A `&self` contract message wrapper.
-#[derive(Debug, Default)]
-pub struct MsgRef<M>
-where
-    M: MessageRef,
-{
-    message: Placeholder<M>,
+/// Connector trait: Connects enum dispatcher for constructors with the contract.
+pub trait ConstructorDispatcher {
+    /// The contract's constructors dispatcher type.
+    type Type;
 }
 
-impl<M> FnSelector for MsgRef<M>
-where
-    M: MessageRef,
-{
-    const SELECTOR: Selector = <M as FnSelector>::SELECTOR;
+/// Connector trait used to start the execution of a smart contract.
+///
+/// The generated message and constructor dispatch enums implement this trait
+/// in order to forward their already decoded state to the selected messages
+/// or constructors.
+pub trait Execute {
+    /// Starts the smart contract execution.
+    fn execute(self) -> Result<()>;
 }
 
-impl<M> Dispatch for MsgRef<M>
+/// Executes the given `&self` message closure.
+///
+/// # Note
+///
+/// The closure is supposed to already contain all the arguments that the real
+/// message requires and forwards them.
+pub fn execute_message<M, F>(f: F) -> Result<()>
 where
     M: MessageRef,
+    F: FnOnce(&<M as FnState>::State) -> <M as FnOutput>::Output,
 {
-    #[inline(always)]
-    fn dispatch(data: &CallData) -> Result<()> {
-        use scale::Decode as _;
-        let args = <M as FnInput>::Input::decode(&mut &data.params()[..])
-            .map_err(|_| DispatchError::InvalidParameters)?;
-        alloc::initialize(ContractPhase::Call);
-        let root_key = Key::from([0x00; 32]);
-        let state =
-            ManuallyDrop::new(pull_spread_root::<<M as FnState>::State>(&root_key));
-        let result = <M as MessageRef>::CALLABLE(&state, args);
-        alloc::finalize();
-        if TypeId::of::<<M as FnOutput>::Output>() != TypeId::of::<()>() {
-            ink_core::env::output::<<M as FnOutput>::Output>(&result)
-        }
-        Ok(())
+    alloc::initialize(ContractPhase::Call);
+    let root_key = Key::from([0x00; 32]);
+    let state = ManuallyDrop::new(pull_spread_root::<<M as FnState>::State>(&root_key));
```

Comment:
Can you explain the use of `ManuallyDrop` here?

Reply:
When `Drop::drop` is called on a storage entity the storage entity tries to clean-up its state on the storage. By packing it right into a `ManuallyDrop` we prevent it from doing this. Note that we generally want this behavior of immediate clean-up upon `Drop::drop` call, however in this case we are looking at a special case that we explicitely want to avoid this for our static storage fields.